src/instance.rs

```diff
@@ -380,7 +380,7 @@ where
 
 impl<'py, T> AsRef<Bound<'py, PyAny>> for Bound<'py, T>
 where
-    T: AsRef<PyAny>,
+    T: HasPyBaseType,
```

Comment:
One more fundamental question: Do we need a bound here at all? If `Bound::as_any` can be called without this as soon as we have an instance of `Bound<T>` at hand, what does the additional bound here achieve?

Reply:
I believe if we don't put a bound on it we'll get infinite recursion while auto-dereferencing, because it would be also available on `Bound<PyAny>`.

Reply:
But auto-derefercing happens only as often as necessary until a resolution can be found, so why would it dereference again after having reached `Bound<PyAny>` as any method that can be will have been found? And a non-existing method will not exist, but I expect the compiler to have safe guards against infinite recursion, either a hard limit or loop detection.

Or does this not compile at all? Did you try just dropping the bound?

Reply:
```console
error[E0055]: reached the recursion limit while auto-dereferencing `instance::Bound<'_, types::any::PyAny>`
  --> src/types/typeobject.rs:72:14
   |
72 |             .to_owned()
   |              ^^^^^^^^ deref recursion limit reached
   |
   = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`pyo3`)
```

> But auto-derefercing happens only as often as necessary until a resolution can be found, so why would it dereference again after having reached Bound<PyAny> as any method that can be will have been found?

So this appears to be wrong.

> And a non-existing method will not exist, but I expect the compiler to have safe guards against infinite recursion, either a hard limit or loop detection.

And this appears to be a hard limit defaulting to 256 steps.

Reply:
I guess this hits this one: https://github.com/rust-lang/rust/issues/19509

Reply:
So this appears to be required for `Deref`. I would suggest dropping it for `AsRef` tough and making a code comment for the `Deref` impl.

> Should we keep the HasPyBaseType name?

As for the name shed, I would add `CoercePyAny` following std's `CoerceUnsized`.

> Should the trait be public or #[doc(hidden)]

I'd suggest keeping the machinery required to implement native types public/documented as downstream crates might need to wrap other native types than those provided in the pyo3 crate, e.g. `rust-numpy` has `PyArray<T, D>` for which the existing `pyobject_native_type_*` macros proved insufficient to handle.

Reply:
> So this appears to be required for `Deref`. I would suggest dropping it for `AsRef` tough and making a code comment for the `Deref` impl.

That sounds good to me, I'll update that.

> As for the name shed, I would add `CoercePyAny` following std's `CoerceUnsized`.

I think I'd be fine with that, but I want to note here that I would consider coercion a distinct mechanism in Rust, which is not really what's happening here. I can see the similarities, but it also could potentially cause confusion. Let's see what other think about that.

> I'd suggest keeping the machinery required to implement native types public/documented as downstream crates might need to wrap other native types than those provided in the pyo3 crate, e.g. `rust-numpy` has `PyArray<T, D>` for which the existing `pyobject_native_type_*` macros proved insufficient to handle.

👍

Reply:
> I think I'd be fine with that, but I want to note here that I would consider coercion a distinct mechanism in Rust, which is not really what's happening here. I can see the similarities, but it also could potentially cause confusion. Let's see what other think about that.

Agreed with this, `CoercePyAny` is reasonable but I'd sort of expect `Bound<T>` to coerce to `Bound<PyAny>` as a non-reference too if it was proper coercion. (e.g. in `fn foo(any: Bound<PyAny>)` pass any `Bound` and get the coercion.)

Some other names I can think of:
- `IsNotPyAny`
- `BoundDerefToPyAny`

Reply:
I would additionally nominate `DerefToPyAny` then as I don't this is limited to `Bound` as the outer wrapper, i.e. we could do this for `Borrowed<T>` as well (if it did not already `Deref` to `Bound<T>`).

(I dislike `IsNotPyAny` because traits should IMHO encapsulate the required properties and/or operations, not type identities. But then again, the whole bound seems to work around what I'd consider a quality of implementation issue in rustc as linked above.)

Reply:
`DerefToPyAny` seems clear enough 👍